auth: update the project authentication method to JWT

marcelosalloum opened the pull request:
### What

This PR enhances the authentication mechanism of the wallet-backend by utilizing JSON Web Tokens (JWT) with Ed25519 signatures. The JWTs are signed using an Ed25519 private key derived from a Stellar secret seed. The server can be configured to accept multiple public keys through the `CLIENT_AUTH_PUBLIC_KEYS` environment variable.

The JWT payload field should contain the following fields:

- (default) `exp` – The expiration time on and after which the JWT must not be accepted for processing, in seconds since Epoch. (Must be less than `iat`+15sec.)
- (default) `iat` - The time at which the JWT was issued, in seconds since Epoch.
- (default) `aud` – The audience for the JWT. This is the server's hostname.
- (default) `sub` – The subject of the JWT, which is the public key of the Stellar account that is being authenticated.
- (custom) `methodAndPath` – The HTTP method and path of the request (e.g., `GET /transactions/b9d0b2292c4e09e8eb22d036171491e87b8d2086bf8b265874c8d182cb9c9020`).
- (custom) `bodyHash`, a hex-encoded SHA-256 hash of the raw HTTP request body, present even when the body is empty:
    ```go
    func HashBody(body []byte) string {
       hashedBodyBytes := sha256.Sum256(body)
       return hex.EncodeToString(hashedBodyBytes[:])
    }
   ```

### Why

The previous signature method was not an industry-standard.

### Issue that this PR addresses

Closes #151 

### Checklist

#### PR Structure

- [x] It is not possible to break this PR down into smaller PRs.
- [x] This PR does not mix refactoring changes with feature changes.
- [x] This PR's title starts with name of package that is most changed in the PR, or `all` if the changes are broad or impact many packages.

#### Thoroughness

- [x] This PR adds tests for the new functionality or fixes.
- [x] All updated queries have been tested (refer to [this](https://stackoverflow.com/a/29163465) check if the data set returned by the updated query is expected to be same as the original one).

#### Release

- [x] This is not a breaking change.
- [x] This is ready to be tested in development.
- [x] The new functionality is gated with a feature flag if this is not ready for production.

marcelosalloum commented on pkg/wbclient/auth/jwt_manager.go:
```diff
@@ -0,0 +1,188 @@
+	_, ed25519PrivateKey, err := ed25519.GenerateKey(bytes.NewReader(privateKeyBytes))
+	if err != nil {
+		return "", fmt.Errorf("generating ed25519 key pair: %w", err)
+	}
```
https://github.com/stellar/wallet-backend/pull/172/commits/402409c2d1e212246cf71ed06744a5e2df4605e1

marcelosalloum commented on pkg/wbclient/auth.go:
```diff
@@ -16,35 +18,21 @@ type SignerInterface interface {
+// SignHTTPRequest adds the generated JWT token to the HTTP request's Authorization header.
+func (sc *RequestSigner) SignHTTPRequest(req *http.Request, timeout time.Duration) error {
 	// Ensure the hostname is parsed correctly
+	hostname := req.URL.Hostname()
 	if !strings.Contains(hostname, "://") {
 		hostname = "https://" + hostname
 	}
```
The protocol is not being used actually... the code just injects it so we can have a valid host at the end.

JakeUrban reviewed (commented):
This looks great, although I have a few suggestions / questions.

- Should we include a `iat` claim in addition to `exp`?
- Should we add a `sub` claim so the client can identify itself? The public key could make sense here. This would allow us to verify the JWT against that public key specifically rather than iterating over all configured public keys for each request.
- Do we want to validate the request path and/or method as well? The current JWT is validated to ensure the request body matches the JWT but it could be made to a different endpoint or use a different method

For reference I'd look at how Fireblocks does authentication (its similar to what you did): https://developers.fireblocks.com/reference/signing-a-request-jwt-structure

aristidesstaffieri commented on internal/serve/middleware/middleware.go:
```diff
@@ -6,50 +6,72 @@ import (
+
+			authHeader := req.Header.Get("Authorization")
+			if authHeader == "" {
+				httperror.Unauthorized("", nil).Render(rw)
```
Should we include the reason in the response here?
`httperror.Unauthorized("missing Authorization header", err).Render(rw)`

marcelosalloum replied:
Thanks for the suggestion! 

I agree it’s helpful during development, but in this case, I’d lean toward not including specific reasons in the response body for security reasons. In general, auth implementations will return limited information to prevent sharing valuable information that would aid bad actors.

The errors are being logged though, so someone with access to the server will be able to debug the issue.

marcelosalloum commented on pkg/wbclient/auth/helpers.go:
```diff
@@ -0,0 +1,37 @@
+		return 0, false
+	}
+
+	return expirationDuration, true
```
https://github.com/stellar/wallet-backend/pull/172/commits/218c8f8a85e7d601f13fe424fcf185600125daf5

marcelosalloum commented on pkg/wbclient/auth/jwt_http_signer_verifier.go:
```diff
@@ -0,0 +1,121 @@
+	tokenString := authHeader[len("Bearer "):] // Remove "Bearer " prefix
+	_, _, err := s.parser.ParseJWT(tokenString, audience, req.URL.Path, bodyBytes)
+	if err != nil {
+		return fmt.Errorf("verifying JWT: %w: %w", err, ErrUnauthorized)
```
The [Go 1.20 Release Notes](https://tip.golang.org/doc/go1.20) state: ￼

The fmt.Errorf function now supports multiple occurrences of the %w format verb, which will cause it to return an error that wraps all of those error operands. ￼

So this approach is totally fine,

marcelosalloum commented:
> This looks great, although I have a few suggestions / questions.
> 
> * Should we include a `iat` claim in addition to `exp`?
> * Should we add a `sub` claim so the client can identify itself? The public key could make sense here. This would allow us to verify the JWT against that public key specifically rather than iterating over all configured public keys for each request.
> * Do we want to validate the request path and/or method as well? The current JWT is validated to ensure the request body matches the JWT but it could be made to a different endpoint or use a different method
> 
> For reference I'd look at how Fireblocks does authentication (its similar to what you did): https://developers.fireblocks.com/reference/signing-a-request-jwt-structure

Hey @JakeUrban, I've addressed your suggestions.
✅ `iat` is now required as well, and we test the it and its relation with `exp` similarly to what Fireblocks do.
✅ The idea of `sub` is really good and it allows us to validate the claims even before validating their signatures. That was incorporated to the multi-jwt-parser as well.
✅ Introduced the `methodAndPath` claim as well.

JakeUrban reviewed (approved):
From a functionality perspective this looks good, thanks for incorporating those suggestions.
